common/lockless_queue.c

```diff
@@ -0,0 +1,143 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. */
+
+#include <openenclave/internal/lockless_queue.h>
+#include <stdlib.h>
+
+#if _MSC_VER
+#include <intrin.h>
+#endif
+
+/* functions for oe_lockless_queue_node */
+/*---------------------------------------------------------------------------*/
+void oe_lockless_queue_node_init(oe_lockless_queue_node* p_node)
+{
+    p_node->p_link = NULL;
+} /* init_oe_lockless_queue_node */
+
+/* functions for oe_lockless_queue */
+/*---------------------------------------------------------------------------*/
+void oe_lockless_queue_init(oe_lockless_queue* p_queue)
+{
+#ifdef _MSC_VER
+    _InterlockedExchangePointer(&(p_queue->p_tail), NULL);
+    _InterlockedExchangePointer(&(p_queue->p_head), NULL);
+#elif defined __GNUC__
+    __atomic_store_n(&(p_queue->p_tail), NULL, __ATOMIC_RELAXED);
+    __atomic_store_n(&(p_queue->p_head), NULL, __ATOMIC_RELAXED);
+#endif /* __GNUC__ */
+} /* init_oe_lockless_queue */
+
+void oe_lockless_queue_push_back(
+    oe_lockless_queue* p_queue,
+    oe_lockless_queue_node* p_node)
+{
+    oe_lockless_queue_node* p_expected = NULL;
+#ifdef _MSC_VER
+    oe_lockless_queue_node* p_actual = NULL;
+#endif /* _MSC_VER */
+    do
+    {
+#ifdef _MSC_VER
+        p_expected = p_actual;
+        p_node->p_link = p_expected;
+        p_actual = (oe_lockless_queue_node*)_InterlockedCompareExchangePointer(
+            &(p_queue->p_tail), p_node, p_expected);
+    } while (p_actual != p_expected);
+#elif defined __GNUC__
+        p_node->p_link = p_expected;
+    } while (!__atomic_compare_exchange_n(
+        &(p_queue->p_tail),
+        &p_expected,
```

Comment:
I don't follow this logic. 
I would expect that a successful push_back operation would have the following post-conditions:

1. p_node->p_link == NULL
2. p_queue->p_tail == p_node

Since __atomic_compare_exchange_n modifies p_expected in case of failure, it seems like we are constructing a (reverse) linked list starting at tail.

Is that the case?

Reply:
Short summary:  The implementation of a lockless data structure is limited by the constraints of atomic operations.  The data structure needs to always be in a consistent stable state across threads.  The only guarantee is that individual atomic operations are consistent between threads.  There is no way to assure that multiple atomic operations happen without thread context changes.
Due to these constraints, this FIFO queue is implemented as two stacks.  This queue implementation is thread safe for multiple producers and a single consumer.
p_tail is the "push" stack.  The implementation of oe_lockless_queue_push_back enables multiple threads to call push_back concurrently.  The compare-exchange operator ensures that conflicts between threads are resolved in a manner that ensures the state of the data structure is stable.
p_head is the "pop" stack.  The implementation of oe_lockless_queue_pop_back is only safe for one thread to call concurrently, but it is safe to call concurrently with push_back. This method first checks to see if there are any nodes on the pop stack and if so it pops a node off and returns.  In the case that there aren't any nodes on the pop stack, it pops all of the nodes off of the push stack, reverses the order, and returns the new node at the top of the stack (front of the queue).

Reply:
@EMumau Thanks for the explanation. Why not add these design/implementation decisions as comments in the code in the first place. Otherwise, the reader is left with deciphering the implementation details/guarantees.

Reply:
Based on your description, this queue handles _multiple-producers, single-consumer_ scenario (the pop can only be safely called only by one thread).
This is bit surprising for use in implementing switchless calls. How are the queues going to be used for implementing ecalls? Is it going to be 1 queue for each enclave worker thread?

Reply:
I have added comments to lockless_queue.h to clarify the usage.

To use this implementation of a lockless queue with *multiple-consumers* there would need to be some additional locking mechanism (i.e. spinlock) around the call to oe_lockless_queue_pop_front.  Every implementation of thread-safe queues that I have seen has been *single-consumer*.